File: Minimum_Difficulty_of_a_Job_Schedule.cpp
// Title: 1335. Minimum Difficulty of a Job Schedule
// Difficulty: Hard
// Problem: https://leetcode.com/problems/minimum-difficulty-of-a-job-schedule/description

class Solution {
public:
    int minDifficulty(std::vector<int>& jobDifficulty, int d) {
        int n = jobDifficulty.size();
        if (n < d) {
            return -1;
        }

        std::vector<std::vector<int>> memo(n, std::vector<int>(d, -1)); // Memoization array

        return dfs(jobDifficulty, d, 0, memo);
    }

private:
    int dfs(const std::vector<int>& jobDifficulty, int days, int index, std::vector<std::vector<int>>& memo) {
        int n = jobDifficulty.size();
        if (index == n && days == 0) {
            return 0;
        }
        if (index == n || days == 0) {
            return INT_MAX;
        }
        if (memo[index][days - 1] != -1) {
            return memo[index][days - 1];
        }

        int maxDifficulty = 0;
        int minDifficulty = INT_MAX;
        for (int i = index; i < n; ++i) {
            maxDifficulty = std::max(maxDifficulty, jobDifficulty[i]);
            int remainingDays = dfs(jobDifficulty, days - 1, i + 1, memo);
            if (remainingDays != INT_MAX) {
                minDifficulty = std::min(minDifficulty, maxDifficulty + remainingDays);
            }
        }

        memo[index][days - 1] = minDifficulty;
        return minDifficulty;
    }
};

File: Minimum_Difficulty_of_a_Job_Schedule.py
# TItle: 1335. Minimum Difficulty of a Job Schedule
# Difficulty: Hard
# Problem: https://leetcode.com/problems/minimum-difficulty-of-a-job-schedule/description/

class Solution:
    def minDifficulty(self, jobDifficulty: List[int], d: int) -> int:
        ## solution using DP, using cache decorator
        n = len(jobDifficulty)
        # not enough jobs for each day
        if n < d:
            return -1
        
        @functools.lru_cache(None)
        def dp(jobs_left: int, days_left: int):
            if days_left == 1:
                return max(jobDifficulty[-jobs_left:])
            difficulty = jobDifficulty[-jobs_left]
            jobs_left -= 1
            days_left -= 1
            # total minumum difficulty for last days_left days
            total_difficulty = difficulty + dp(jobs_left, days_left)

            # increase the number of difficult tasks for today
            while jobs_left > days_left:
                difficulty = max(difficulty, jobDifficulty[-jobs_left])
                jobs_left -= 1
                total_difficulty = min(total_difficulty, difficulty + dp(jobs_left, days_left))

            return total_difficulty

        return dp(n, d)
